### ISSUE TITLE
Module naming scheme `OPTIONS` and the `emacs` package

### ISSUE THREAD

**Issue by TheQueasle** [LLNL/spack#2884]
Playing around with the `modules.yaml` feature and setting
```
naming_scheme: '${PACKAGE}/${PACKAGE}-${VERSION}-${OPTIONS}-${COMPILERNAME}-${COMPILERVER}'
```

It leads to `emacs` with the following name `emacs/emacs-25.1?toolkit=gtk?-gcc-4.8.5`:
```
[vagrant@sandbox ~]$ module avail emacs

-------------- /vagrant/spack/share/spack/modules/linux-centos7-x86_64 --------------
emacs/emacs-25.1?toolkit=gtk?-gcc-4.8.5
[vagrant@sandbox ~]$ cd /vagrant/spack/share/spack/modules/linux-centos7-x86_64/emacs/
[vagrant@sandbox emacs]$ ls -l
total 4
-rw-r--r-- 1 vagrant vagrant 974 Jan 23 10:22 emacs-25.1 toolkit=gtk -gcc-4.8.5
```

Very quick guess, it's because `spec.variants['toolkit'].value` is one of `gtk` or `athena` and not `True` or `False`?
```
        toolkit = spec.variants['toolkit'].value
        if '+X' in spec:
            if toolkit not in ('gtk', 'athena'):
                raise InstallError("toolkit must be in (gtk, athena), not %s" %
                                   toolkit)
            args = [
                '--with-x',
                '--with-x-toolkit={0}'.format(toolkit)
            ]
        else:
            args = ['--without-x']
```

**Comment by TheQueasle** [comment]
Oh!  This is in the **develop** branch, btw...

**Comment by TheQueasle** [comment]
I think this is closely related to the topics discussed in the PR #2386

Multi-valued variants and the module file naming scheme logic for `OPTIONS` kind of clash in the current state of develop.

**Comment by adamjstewart** [comment]
@alalazo

**Comment by alalazo** [comment]
@TheQueasle Just to give some context: in #1848 the substitution scheme for module files was changed from an ad-hoc set of keyword substitutions to the use of `spec.format`. The rationale was that we will merge features making heavy use of that method for formatting and the responsibilities of the two forms of substitution were clearly overlapping. What we didn't do was checking that every new keyword would make sense / behave well in the context of modules. `OPTIONS` for instance is a keyword that was not present among the ad-hoc ones previously used. 

@tgamblin @scheibelp and everybody here: should we limit the set of substitutions that are viable in this context or should we fix the formatting above? I would vote for the former, as usually module files are used to present a simplified view of what is installed and there are other mechanisms to append suffixes to module file names.

**Comment by TheQueasle** [comment]
@alalazo Thanks for the context!  So, just to make sure I'm following along correctly, `OPTIONS` did not exist prior to #1848. The substitution scheme now uses `spec.format` and `OPTIONS` is translated into whatever `spec.format('$+')` resolves to, which in this case is the string ' toolkit=gtk '.  (Sorry if I just restated the obvious.  I'm learnin' :school: :smile: )

When you say 'should we limit the set of substitutions that are viable' what do you mean?  Does that mean limiting or doing away with the use of `OPTIONS` or a like scheme?

**Comment by alalazo** [comment]
@TheQueasle @tgamblin The tokens that can be substituted now in the naming scheme are:
```
${PACKAGE}       Package name
${VERSION}       Version
${COMPILER}      Full compiler string
${COMPILERNAME}  Compiler name
${COMPILERVER}   Compiler version
${COMPILERFLAGS} Compiler flags
${OPTIONS}       Options
${ARCHITECTURE}  Architecture
${SHA1}          Dependencies 8-char sha1 prefix
${HASH:len}      DAG hash with optional length specifier
${SPACK_ROOT}    The spack root directory
${SPACK_INSTALL} The default spack install directory,
${PREFIX}        The package prefix
```
In #3183 I would like to limit them to:
```
${PACKAGE}       Package name
${VERSION}       Version
${COMPILER}      Full compiler string
${COMPILERNAME}  Compiler name
${COMPILERVER}   Compiler version
```
would it be fine with you?

**Comment by TheQueasle** [comment]
@alalazo I'm not sure.  Maybe I'm not understanding how #3183 works.  Without the hash, how do we avoid filename collisions on the module files generated?

Sorry if that is a silly question.  I've not had time yet to really dive into that PR and play.

I'm also currently using a modified version of OPTIONS to only include the enabled variants in a modulefile name.  I can describe that more if need be.

**Comment by alalazo** [comment]
@TheQueasle The hash has always been appended automatically at the end of the filename and don't need to appear in the naming scheme. If you look at the current `develop` in `modules.py` you'll see the method:
```python
def _get_suffixes(self):
    configuration, _ = parse_config_options(self)
    suffixes = []
    for constraint, suffix in configuration.get('suffixes', {}).items():
        if constraint in self.spec:
            suffixes.append(suffix)
    hash_length = configuration.get('hash_length', 7)
    if hash_length != 0:
        suffixes.append(self.spec.dag_hash(length=hash_length))
    return suffixes
```
This method gets called when you construct the `use_name` for a given module system. You have control over the length of the hash (that by default is 7) using the configuration option `hash_length`, see the tutorial [here](https://spack.readthedocs.io/en/latest/tutorial_sc16_modules.html#change-module-file-naming). In that section of the tutorial it is also shown how to map any anonymous spec to a suffix that will be added to the name.

**Comment by alalazo** [comment]
> I'm also currently using a modified version of OPTIONS to only include the enabled variants in a modulefile name. I can describe that more if need be.

I think this will be covered by the `suffixes` keyword that you can already use in `modules.yaml`. Actually  you can even say thing like:
```yaml
netlib-scalapack:
  suffixes:
    '^openblas+openmp': 'openmp'
```
which means that the module file name for `netlib-scalapack` will have an `openmp` suffix if it depends on `openblas+openmp`. Am I right or am I overlooking some cases?

**Comment by scheibelp** [comment]
> I think this will be covered by the suffixes keyword that you can already use in modules.yaml. Actually you can even say things like

@alalazo I presume this would require a separate entry for each enabled option? I'm not opposed to restricting the tokens (although I'm having trouble thinking of a way to constrain spec.format in that manner straightforwardly) although if @TheQueasle adds a spec.format entry like {ENABLED_OPTIONS}  (which IMO would be nice) then perhaps that could be allowed too?

**Comment by alalazo** [comment]
> @alalazo I presume this would require a separate entry for each enabled option? I'm not opposed to restricting the tokens (although I'm having trouble thinking of a way to constrain spec.format in that manner straightforwardly) although if @TheQueasle adds a spec.format entry like {ENABLED_OPTIONS} (which IMO would be nice) then perhaps that could be allowed too?

I don't think it's `spec.format` that needs to be constrained, but the module system that needs to inspect the naming scheme before formatting it. 

For `{ENABLED_OPTIONS}` I don't have a clear opinion. I am not opposed to add it (and I think it would be good in some cases), but I also know that if I am going to install `boost` then the file name will be way too long.

**Comment by TheQueasle** [comment]
@alalazo - Yeah...the filename length is something I was attempting to combat with `{ENABLED_OPTIONS}`, but you are right, it won't save me from packages that have a lot of enabled options...

> I think this will be covered by the suffixes keyword that you can already use in modules.yaml. 

This could work.  It's not as 'automagic' as my laziness would hope for...as it seems I would need to define my suffixes by hand for packages that are installed multiple times with only differing enabled variants.

**Comment by becker33** [comment]
@alalazo I think that `${ARCHITECTURE}` should be included in the tokens we can use for modules. I think everything else may be better specified on a per-package basis (which would include dependencies as well).

**Comment by alalazo** [comment]
@becker33 In the end in #3183 I am not limiting the tokens that can be substituted. I would prefer to submit small improvements like this one as a separate PR after #3183 will be merged.
